src/player.cpp

```diff
@@ -4362,10 +4373,39 @@ void player::update_stamina( int turns )
         }
     }
 
+    // Decrease stamina regeneration if tired
+    float fatigue_mod = std::max( ( 1 + get_fatigue() / 1000.0f ), 0.1f );
+    stamina_recovery *= fatigue_mod;
+
     stamina = roll_remainder( stamina + stamina_recovery * turns );
 
     // Cap at max
-    stamina = std::min( std::max( stamina, 0 ), max_stam );
+    const int max_stam_penalty = get_stamina_max() - get_stamina_max_penalty();
+    stamina = std::min( std::max( stamina, 0 ), max_stam_penalty );
+
+    // Add winded effect
+    if( ( stamina < ( get_stamina_max() / 10 ) ) && !has_effect( effect_winded ) ){
+        add_effect( effect_winded, 10_turns );
+    }
+
+    // increase needs with stamina recovery
+    int StaminaIncreaseHunger = get_option< int >( "PLAYER_RECOVER_STAMINA_INCREASE_HUNGER" );
+    float AddHunger = roll_remainder( stamina_recovery * turns / 1000.0f * StaminaIncreaseHunger );
+    if( is_npc() ) {
+        AddHunger *= 0.25f;
+    }
+    mod_stat( "hunger", AddHunger );
+
+    int StaminaIncreaseThrist = get_option< int >( "PLAYER_RECOVER_STAMINA_INCREASE_THRIST" );
+    float AddThrist = roll_remainder( stamina_recovery * turns / 1000.0f * StaminaIncreaseThrist );
+    if( is_npc() ) {
+        AddThrist *= 0.25f;
+    }
+    mod_stat( "thirst", AddThrist );
+
+    int StaminaIncreaseFatigue = get_option< int >( "PLAYER_RECOVER_STAMINA_INCREASE_FATIGUE" );
+    float AddFatigie = roll_remainder( stamina_recovery * turns / 1000.0f * StaminaIncreaseFatigue );
```

Comment:
Typo: should be "AddFatigue".
Also, no fatigue increase for npc?

Reply:
Should npc have 1/4 rate of stamina decreasing, like for hunger and thrist? They must have buff, but not sure how to implement it better.